### PR TITLE
add rust cfg file to kbuild

### LINKED ISSUE #22: Avoid using a bunch of cfg flags and use of hardcoded auto.conf path
Currently, building with O=$something fails as auto.conf is search for in the source tree.
Also the arguments passed as cfg parsed from auto.conf can get bigger as the max argument length of linux.

### PR DESCRIPTION
**Pull request by kloenk**
Fixes #22

Contains #28

### CODE REVIEW

**Comment by kloenk** [comment]
I created a file containing --cfg arguments via the kbuild c script.
rustc can take this file if we give it via @<filename>. We just have to figure out the absolute (or relative path from where rustc is invoked) to that file an append it to the rustc arguments

**Comment by kloenk** [comment]
@ojeda 
The ci does seem to fail because of other failures (404?) maybe someone can restart the job if we are lucky this helps?

**Comment by ojeda** [comment]
Yeah, something odd is going on in Azure (re-ran them now). Let me invite you to the organization so that you can re-run them too.

**Comment by kloenk** [comment]
This somehow expanded to get it to build in general, so could take some time

**Comment by kloenk** [comment]
I somehow destroyed the symbols. Not yet sure how.

**Comment by kloenk** [comment]
Seems like that for some reasons make does not pass --cfg=MODULE to rustc, not sure why.

Does somebody with more makefile expertise want to look at it?

**Comment by kloenk** [comment]
Seems like it works now.

**Comment by kloenk** [comment]
Now builtin fails

**Comment by ojeda** [comment]
> Now builtin fails

It might be the #20 issue that re-surfaces again here (I haven't taken a look at solving that panic yet, sorry).

**Comment by kloenk** [comment]
I think it's something else. The panic from #20 prints output from rust. here it panics before printing module information.

It seems like this is the problem (unfortunately I don't know yet how to debug the kerne, but will try to learn)
~~~
[    0.567597] BUG: kernel NULL pointer dereference, address: 0000000000000004

**Comment by kloenk** [comment]
Sadly I cannot build the kernel with debug symbols, which make it harder to debug. Not sure why it fails, maybe some broken ld or so.

**Comment by kloenk** [comment]
This also failes in `v5.10-rc2`.

Do you @ojeda know on which list or bugzilla I should report that bug? linux-kbuild or linux-kernel?

**Comment by ojeda** [comment]
> This also failes in `v5.10-rc2`.
> 
> Do you @ojeda know on which list or bugzilla I should report that bug? linux-kbuild or linux-kernel?

We would need to reproduce it without all our other code first, i.e. in the mainline kernel.

> Sadly I cannot build the kernel with debug symbols, which make it harder to debug. Not sure why it fails, maybe some broken ld or so.

What do you mean? `ld` gives an error?

**Comment by kloenk** [comment]
> We would need to reproduce it without all our other code first, i.e. in the mainline kernel.

Took a clean [`v5.10-rc2`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=3cea11cd5e3b00d91caf0b4730194039b45c5891) and build from there.


> What do you mean? `ld` gives an error?

I see this error. Not sure if my ld is broken, or what the problem is.
This only occurred for me when building with `DEBUG y`.

~~~
  CHK     include/generated/compile.h
  LD [M]  fs/efivarfs/efivarfs.o
/nix/store/p792j5f44l3f0xi7ai5jllwnxqwnka88-binutils-2.31.1/bin/ld: fs/efivarfs/inode.o: unable to initialize decompress status for section .debug_loc
/nix/store/p792j5f44l3f0xi7ai5jllwnxqwnka88-binutils-2.31.1/bin/ld: fs/efivarfs/inode.o: unable to initialize decompress status for section .debug_loc
/nix/store/p792j5f44l3f0xi7ai5jllwnxqwnka88-binutils-2.31.1/bin/ld: fs/efivarfs/inode.o: unable to initialize decompress status for section .debug_loc
/nix/store/p792j5f44l3f0xi7ai5jllwnxqwnka88-binutils-2.31.1/bin/ld: fs/efivarfs/inode.o: unable to initialize decompress status for section .debug_loc
fs/efivarfs/inode.o: file not recognized: file format not recognized
make[3]: *** [../scripts/Makefile.build:434: fs/efivarfs/efivarfs.o] Error 1
make[2]: *** [../scripts/Makefile.build:500: fs/efivarfs] Error 2
make[1]: *** [/persist/data/kloenk/proj/linux/Makefile:1799: fs] Error 2
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/persist/data/kloenk/proj/linux/build'
make: *** [Makefile:185: __sub-make] Error 2
~~~

(The /nix/store path is just where ld is in my path)

**Comment by ojeda** [comment]
> Took a clean [`v5.10-rc2`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=3cea11cd5e3b00d91caf0b4730194039b45c5891) and build from there.

Ah, so this is another thing? To clarify, we have 3 issues now then?
  - The #20 panic
  - The panic (here)
  - The build failure (here)

> I see this error. Not sure if my ld is broken, or what the problem is.
> This only occurred for me when building with `DEBUG y`.
> 
> ```
> /nix/store/p792j5f44l3f0xi7ai5jllwnxqwnka88-binutils-2.31.1/bin/ld: fs/efivarfs/inode.o: unable to initialize decompress status for section .debug_loc
> ```

That looks like https://wiki.gentoo.org/wiki/Binutils_2.32_upgrade_notes/elfutils_0.175:_unable_to_initialize_decompress_status_for_section_.debug_info (and similar reports). Can you please try with a newer binutils? (or with LLVM -- if you have version 10.0.1 at least, try passing `LLVM=1` to make)

**Comment by kloenk** [comment]
> > Took a clean [`v5.10-rc2`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=3cea11cd5e3b00d91caf0b4730194039b45c5891) and build from there.
> 
> Ah, so this is another thing? To clarify, we have 3 issues now then?
>   - The #20 panic
>   - The panic (here)
>   - The build failure (here)

Yes, I see three errors.

> > I see this error. Not sure if my ld is broken, or what the problem is.
> > This only occurred for me when building with `DEBUG y`.
> > 
> > ```
> > /nix/store/p792j5f44l3f0xi7ai5jllwnxqwnka88-binutils-2.31.1/bin/ld: fs/efivarfs/inode.o: unable to initialize decompress status for section .debug_loc
> > ```
> 
> That looks like https://wiki.gentoo.org/wiki/Binutils_2.32_upgrade_notes/elfutils_0.175:_unable_to_initialize_decompress_status_for_section_.debug_info (and similar reports). Can you please try with a newer binutils? (or with LLVM -- if you have version 10.0.1 at least, try passing `LLVM=1` to make)

I'm using LLVM 11. Will try later with LLVM=1, currently only used CC=clang. Newer binutils was not available the last time I checked. Will also look later.

**Comment by ojeda** [comment]
> I'm using LLVM 11. Will try later with LLVM=1, currently only used CC=clang. Newer binutils was not available the last time I checked. Will also look later.

Thanks! Yeah, `CC=clang` is only for the C compiler. `LLVM=1` is a shorthand for `CC=clang` plus the rest of the toolchain, including lld etc.

**Comment by kloenk** [comment]
I don't have enough time at hand to learn kernel debugging. Has anyone time to debug this error? or do we want to disable the builtins test for now, and say that it cannot be a builtin kernel module and fix it later?

**Comment by kloenk** [comment]
#28 does not fix this kernel panic. So could not test #28 yet.

**Comment by ojeda** [comment]
Thanks for trying! We could have been lucky. I will take a look at this one then.

**Comment by kloenk** [comment]
I saw a pagefault. The stacktrace contains something about unit, but I saw no errors regarding do_one_initcall

**Comment by kloenk** [comment]
pushed, so you have my current local state.

File: scripts/kconfig/confdata.c [hunk 1]
```diff
@@ -636,6 +636,55 @@ static struct conf_printer kconfig_printer_cb =
...
+		break;
+	}
+
+	fprintf(fp, "--cfg=%s%s=%s\n", CONFIG_, sym->name, str);
+}
+
+static void
+rconfig_print_comment(FILE *fp, const char *value, void *arg)
+{
+	fprintf(stderr, "could not print commend\n");
```
Comment (ojeda) [inline scripts/kconfig/confdata.c:678]: Typo

File: scripts/kconfig/confdata.c [hunk 2]
```diff
@@ -636,6 +636,55 @@ static struct conf_printer kconfig_printer_cb =
...
+ * Rust configuration printer
+ *
+ * This printer is used when generating the resulting rustc configuration
+ * after kconfig invocation and `defconfig` files.
+ *
+ */
+static void
+rconfig_print_symbol(FILE *fp, struct symbol *sym, const char *value, void *arg)
+{
+  const char *str;
```
Comment (ojeda) [inline scripts/kconfig/confdata.c:649]: Indentation.

File: scripts/kconfig/confdata.c [hunk 3]
```diff
@@ -636,6 +636,55 @@ static struct conf_printer kconfig_printer_cb =
...
 };
 
+/*
+ * Rust configuration printer
+ *
+ * This printer is used when generating the resulting rustc configuration
+ * after kconfig invocation and `defconfig` files.
+ *
+ */
+static void
```
Comment (ojeda) [inline scripts/kconfig/confdata.c:646]: Please follow the same convention as in the rest of the file (same line for the signature).

File: scripts/ld-version.sh [hunk 4]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/awk -f
+#!/nix/store/4z79ipgxqn80ns7mpax25zmb77i3ndfw-gawk-5.1.0/bin/gawk -f
```
Comment (ojeda) [inline scripts/ld-version.sh]: Spurious change.

File: gawk [hunk 5]
```diff
@@ -0,0 +1 @@
+/nix/store/4z79ipgxqn80ns7mpax25zmb77i3ndfw-gawk-5.1.0
```
Comment (ojeda) [inline gawk]: Spurious change (also the other `gawk*`).

**Comment by ojeda** [comment]
> pushed, so you have my current local state.

Thanks a lot -- I was about to ask how to reproduce it. So these changes should be enough?

File: gawk (hunk 5 above)
Comment (kloenk) [reply gawk]: accidently commited it. should be fixed now.

File: scripts/kconfig/confdata.c (hunk 2 above)
Comment (kloenk) [reply scripts/kconfig/confdata.c:649]: Thanks, will fix later

**Comment by kloenk** [comment]
Yes, I only have those awk patches added to my local version, so I can build it under nixos

File: scripts/kconfig/confdata.c (hunk 3 above)
Comment (ojeda) [reply scripts/kconfig/confdata.c:646]: I saw the file has different styles, so it does not matter.

**Comment by ojeda** [comment]
OK, this one was easy. It is not a problem with the current HEAD. Instead, things got broken here because we didn't have the "naked" `CONFIG_X`s I mentioned in #22, so the `module!` macro was not seeing `CONFIG_HAVE_ARCH_PREL32_RELOCATIONS` and therefore it was breaking the initcall table.

I took the liberty to reduce the differences w.r.t. the current HEAD, so I am sending a PR (#32) with you as author, but already having this fixed. Same for the bindgen proposal part, since you already had it done (https://github.com/Rust-for-Linux/linux/commit/b91aff3b53979d6bdc3fefe6e81dadfd3ac9260c). With those two things out of the way, then the only differences that remain here are your `O=...` fixes.

**Comment by kloenk** [comment]
Will reopen as new pr, with only the O= thing

**Pull request closed without merging**